packages/lint-eslint-config-rules/src/index.ts

```diff
@@ -0,0 +1,155 @@
+// eslint-disable-next-line import/extensions
+import 'core-js/es/set/index.js'; // for Node.js <22.0.0
+
+import { basename } from 'node:path';
+
+import { FlatCompat } from '@eslint/eslintrc';
+import * as fsWalk from '@nodelib/fs.walk';
+import { ESLint, type Rule } from 'eslint';
+import { builtinRules } from 'eslint/use-at-your-own-risk';
+
+const defaultCwd = process.cwd();
+
+/** `${ruleId}` (from ESLint core) or `${pluginName}/${ruleName}` (from ESLint plugin) */
+export type ESLintRuleId = string;
+export interface IESLintRule {
+  id: ESLintRuleId;
+  meta: Rule.RuleMetaData | undefined;
+}
+export type ESLintRuleMap = Map<ESLintRuleId, IESLintRule>;
+
+export interface IESLintConfigRulesLintResult {
+  // other context
+  readonly eslint: ESLint;
+  readonly compat: FlatCompat;
+
+  readonly pluginMap: Map<string, ESLint.Plugin>;
+  /** all rules that can be used */
+  readonly ruleMap: ESLintRuleMap;
+
+  // calculated result
+  /**
+   * Rules that specified in the config file
+   *
+   * Whether the rule is 'off' or 'warn' or 'error' is not considered.
+   */
+  readonly usedRuleIds: Set<string>;
+  readonly usedPluginSpecifiers: Set<string>;
+
+  readonly usedUnknownRuleIds: Set<string>;
+  readonly usedKnownRuleIds: Set<string>;
+  readonly usedDeprecatedRuleIds: Set<string>;
+}
+
+const collator = new Intl.Collator('en');
+/**
+ * Sort ruleIds in a way that
+ * core rules come first, then plugin rules
+ * for human readability
+ */
+export const sortedRuleIds = (ruleIds: Iterable<string>): string[] =>
+  Array.from(ruleIds).sort((a, b) => {
+    if (a.includes('/') && !b.includes('/')) {
+      return 1;
+    }
+    if (!a.includes('/') && b.includes('/')) {
+      return -1;
+    }
+    return collator.compare(a, b);
+  });
+
+export const lintESLintConfigRules = async (
+  /**
+   * The directory to lint, default to `process.cwd()`
+   */
+  cwd = defaultCwd,
+): Promise<IESLintConfigRulesLintResult> => {
+  const eslint = new ESLint({ cwd });
+  const compat = new FlatCompat({
+    baseDirectory: cwd,
+    resolvePluginsRelativeTo: cwd,
+  });
+
+  let usedRuleIds: Set<string> = new Set();
+  let usedPluginSpecifiers: Set<string> = new Set();
+  const rulesMap: ESLintRuleMap = new Map(
+    Array.from(builtinRules.entries(), ([ruleId, rule]) => [
+      ruleId,
+      { id: ruleId, meta: rule.meta } satisfies IESLintRule,
+    ]),
+  );
+
+  /**
+   * Iterate over all files in the project to
+   * collect used rules and plugin specifiers
+   */
+  for (const entry of fsWalk.walkSync(
```

Comment:
Using `fsWalk.walkSync` with async `calculateConfigForFile` makes the loop sequential, impacting performance. Suggest refactoring to asynchronous `fsWalk.walk` with `Promise.all` for parallel execution.

```ts
const entries = await fsWalk.walk(cwd, settings);
const configs = await Promise.all(entries.map(entry => eslint.calculateConfigForFile(entry.path)));
```

Reply:
[`fsWalk.walk` is a callback-style API](https://www.npmjs.com/package/@nodelib/fs.walk#api), and it returns all entries at once, just like `fs.walkSync`. In my opinion, we cannot do much with it.

I've refactored the asynchronous `calculateConfigForFile` calls into a `Promise.all()`

Reply:
I apologize for the confusion in my previous comment. Upon further review of the fs.walk API documentation, I realized I overlooked an important detail. The documentation states: 
>  If you want to use the Promise API, use `util.promisify`.

Here's a corrected version that should work:

```ts
import { promisify } from 'util';

const walkAsync = promisify(fsWalk.walk);

const entries = await walkAsync(cwd, new fsWalk.Settings({
  deepFilter: (_entry) =>
    !['.git', 'node_modules'].includes(basename(_entry.path)),
  entryFilter: (_entry) => _entry.dirent.isFile(),
}));

await Promise.all(
  entries.map(async (entry) => {
    const config = await eslint.calculateConfigForFile(entry.path) as
      | undefined
      | {
          rules: Record<string, unknown>;
          plugins: string[];
        };

    if (config !== undefined) {
      usedRuleIds = usedRuleIds.union(new Set(Object.keys(config.rules)));
      usedPluginSpecifiers = usedPluginSpecifiers.union(
        new Set(config.plugins),
      );
    }
  })
);
```

This approach maintains the benefits of parallel processing that I initially suggested, while correctly handling the callback-style API of `fsWalk.walk`. Do you think this revised approach would be beneficial for the performance of the function?

Reply:
Yes, we can use util.promisify() to convert fsWalk.walk to a promise-style function. However, we are still waiting all entries are returned before we can process them with `calculateConfigForFile`. The key steps are:

1. Get all entries from `fsWalk`, either by `await util.promisify(fsWalk.walk)(cwd)` or `fsWalk.walkSync(cwd)`.
2. Process entries from 1. with `calculateConfigForFile()`

The step 1 is a solid blocking operation, and we can't start step 2 until all entries are returned. Unless `fsWalk` provides a new API that returns its result as an async generator, we can't start processing each entry as soon as it is discovered.

The step 2 is a bunch of async operations, we can use `Promise.all()` to run them concurrently.

---

However, fsWalk provides a `fsWalk.walkStream()` API that returns results as a stream. We may use it to break down step 1 into smaller chunks, and process them concurrently.

Reply:
The new implementation uses [`fsWalk.walkStream()`](https://github.com/nodelib/nodelib/tree/master/packages/fs/fs.walk#walkstreampath-optionsorsettingsi) and [`readable.forEach()`](https://nodejs.org/docs/latest/api/stream.html#readableforeachfn-options) to achieve fully asynchronous file enumeration and processing.

See the diff for more details:
https://github.com/RightCapitalHQ/frontend-style-guide/compare/5b5ab74381500aa74102fa4e2e9b0ca8870fa381..6f419a8f943894042fbc83acd8726a5b2bfecb69#diff-8c33b3645b2b6a257a0a0145134ed5295361df791d34eb6d62205be024ed2b2bL86-R123